### PR TITLE
ADD module mrp_request_workcenter_cycle 12.0

### PR DESCRIPTION
**Pull request by bealdav**
This module :
- [x] compute manufacturing request quantity from workcenter capacity for a product
- [x] create manufacturing order according to selected workcenter

## Settings
Selection_125

## Usage
request

## TODO

- [x] choose a really better name for the module, and a shorter one
- [x] test
- [x] readme

### CODE REVIEW

File: mrp_production_request_qty_from_workcenter/models/product.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    _description = "Quantity that can be produced with a workcenter cycle"
+
+    workcenter_id = fields.Many2one(
+        comodel_name="mrp.workcenter", string="Workcenter", help=""
+    )
+    product_id = fields.Many2one(
+        comodel_name="product.product", string="Product", help=""
+    )
+    product_qty = fields.Float(help="")
+    workcenter_qty = fields.Float(help="")
```
Comment (bealdav) [inline mrp_production_request_qty_from_workcenter/models/product.py:30]: TODO
```suggestion
    workcenter_cycle = fields.Float(help="")
```

File: mrp_production_request_qty_from_workcenter/models/request.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                }
+                qty = 0
+                for qty_by in rec.qty_by_workcenter_ids:
+                    qty += (
+                        qty_in_product.get(qty_by.workcenter_id, 0)
+                        * qty_by.workcenter_qty
+                    )
+                rec.product_qty = qty
+
+    def button_create_mo_for_workcenter(self):
```
Comment (bealdav) [inline mrp_production_request_qty_from_workcenter/models/request.py]:
```suggestion
    def button_create_mo_by_workcenter(self):
```

File: mrp_production_request_qty_from_workcenter/models/request.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    _name = "production.request.workcenter"
+    _description = "Workcenters attached to Production Request"
+
+    request_id = fields.Many2one(
+        comodel_name="mrp.production.request", string="Production Request", help=""
+    )
+    workcenter_id = fields.Many2one(
+        comodel_name="mrp.workcenter", string="Workcenter", help=""
+    )
+    workcenter_qty = fields.Float(help="")
```
Comment (bealdav) [inline mrp_production_request_qty_from_workcenter/models/request.py]:
```suggestion
    workcenter_cycle = fields.Float(help="")
```

File: mrp_production_request_qty_from_workcenter/models/workcenter.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+class MrpWorkcenter(models.Model):
+    _inherit = "mrp.workcenter"
+
+    qty_by_product_ids = fields.One2many(
+        comodel_name="product.workcenter.quantity",
+        inverse_name="workcenter_id",
+        string="Quantity by product",
+        help=""
+    )
```
Comment (bealdav) [inline mrp_production_request_qty_from_workcenter/models/workcenter.py]: not sure this field is usefull

File: mrp_production_request_qty_from_workcenter/views/product.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record model="ir.ui.view" id="product_product_form_view_bom_button">
+        <field name="model">product.product</field>
+        <field name="inherit_id" ref="mrp.product_product_form_view_bom_button"/>
+        <field name="arch" type="xml">
+            <xpath expr="//group[@name='traceability']" position="after">
+                <group name="manufacturing">
+                    <group string="Quantity by workcenter" groups="mrp_production_request.group_mrp_production_request_user">
```
Comment (bealdav) [inline mrp_production_request_qty_from_workcenter/views/product.xml]:
```suggestion
                    <group string="Capacity by workcenter" groups="mrp_production_request.group_mrp_production_request_user">
```

**Comment by bealdav** [comment]
cc @LoisRForgeFlow @hparfr @florian-dacosta  could you have a look please ?

File: mrp_request_qty_by_workcenter/models/request.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    _description = "Workcenters attached to Production Request"
+
+    request_id = fields.Many2one(
+        comodel_name="mrp.production.request", string="Production Request"
+    )
+    workcenter_id = fields.Many2one(comodel_name="mrp.workcenter", string="Workcenter")
+    workcenter_cycle_no = fields.Float(
+        string="Cycle Number",
+        help="A cycle is complete ended process for the machine.\n"
+        "Multiple cycles are several machines of the same kind in parallele "
```
Comment (hparfr) [inline mrp_request_qty_by_workcenter/models/request.py]:
```suggestion
        "Multiple cycles are several machines of the same kind in parallel"
```

**Review by hparfr — APPROVED** [review]
Functional review OK.
It's a great feature when you want to trigger multiple MO at once.

**Comment by LoisRForgeFlow** [comment]
Hi, I have reviewed this locally, the functionality looks good! 

Regarding the name, I agree that is hard to find a good (and short) name, the only one that comes to my mind is `mrp_production_request_workcenter_cycle" but I don't like it either. What I prefer if possible, is to keep the full prefix of `mrp_production_request_*`, I know it will make the name longer but it also makes the module easier to identify. Side note, maybe we could rename mrp_production_request to mrp_request in the future (v14?).

Also, could you take care of travis and runbot?.

Thanks!

**Comment by bealdav** [comment]
Really nice idea to rename mrp_production_request in mrp_request in v14.

Maybe could be defined in v14 migration issue ?

**Review by dora-jurcevic — APPROVED** [review]
Feature tested and working as described

**Review by mourad-ehm — APPROVED** [review]

**Review by marcelsavegnago — APPROVED** [review]

**Comment by bealdav** [comment]
Thanks a lot @dreispt for your review and merge.

Could you have a look to https://github.com/OCA/manufacture/pull/558 please ?

Thanks again

**Pull request merged**